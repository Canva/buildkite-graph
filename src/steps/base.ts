import ow from 'ow';
import 'reflect-metadata';
import { Expose, Exclude, Transform } from 'class-transformer';

export interface BaseStep {}

// see https://github.com/microsoft/TypeScript/issues/22815#issuecomment-375766197
export interface DefaultStep extends BaseStep {}
export abstract class DefaultStep implements BaseStep {
    @Exclude()
    public readonly dependencies: Set<DefaultStep> = new Set();

    dependsOn(step: DefaultStep): this {
        this.dependencies.add(step);
        return this;
    }

    @Exclude()
    public always: boolean = false;

    alwaysExecute() {
        this.always = true;
        return this;
    }
}

@Exclude()
export class BranchLimitedStep extends DefaultStep {
    @Expose({ name: 'branches' })
    @Transform((branches: Set<string>) =>
        branches.size ? [...branches].sort().join(' ') : undefined,
    )
    private _branches: Set<string> = new Set();

    withBranch(pattern: string): this {
        ow(pattern, ow.string.nonEmpty);
        this._branches.add(pattern);
        return this;
    }
}

export class LabeledStep extends BranchLimitedStep {
    private _label?: string;

    @Expose()
    get label() {
        return this._label;
    }

    withLabel(label: string) {
        this._label = label;
        return this;
    }
}

@Exclude()
export abstract class Chainable<T> {
    constructor(protected readonly parent: T) {
        this.parent = parent;
    }
}

export type ExitStatus = number | '*';

export const exitStatusPredicate = ow.any(
    ow.string.equals('*'),
    ow.number.integer,
);
