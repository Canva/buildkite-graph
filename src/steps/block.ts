import { Exclude, Expose, Transform } from 'class-transformer';
import { BranchLimitedStep } from '../base';
import { Fields, FieldsImpl } from './block/fields';

@Exclude()
export class BlockStep extends BranchLimitedStep {
    @Expose({ name: 'block' })
    private readonly title: string;
    @Expose()
    private readonly prompt?: string;
    @Expose()
    @Transform((value: FieldsImpl<any>) =>
        value.hasFields() ? [...value.fields.values()] : undefined,
    )
    public readonly fields: Fields<this> = new FieldsImpl(this);
    constructor(title: string, prompt?: string) {
        super();
        this.title = title;
        this.prompt = prompt;
    }
    toString() {
        return `[block for '${this.title}']`;
    }
}
