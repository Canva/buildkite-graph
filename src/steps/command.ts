import { Exclude, Expose, Transform } from 'class-transformer';
import ow from 'ow';
import 'reflect-metadata';
import { KeyValue, KeyValueImpl, transformKeyValueImpl } from '../key_value';
import {
    Plugin,
    Plugins,
    PluginsImpl,
    transformPlugins,
} from './command/plugins';
import { ExitStatus, exitStatusPredicate, LabeledStep } from '../base';
import { Retry, RetryImpl } from './command/retry';

function assertTimeout(timeout: number): void {
    ow(timeout, ow.number.integerOrInfinite.positive);
}

function assertSkipValue(value: SkipValue): void {
    if (typeof value === 'string') {
        ow(value, ow.string.nonEmpty);
    }
}

type SkipValue = boolean | string;
export type SkipFunction = () => SkipValue;

export class Command {
    constructor(public command: string, public timeout: number = Infinity) {
        ow(command, ow.string);
        assertTimeout(timeout);
    }

    toString(): string {
        return this.command;
    }
}

type Agents = Map<string, string>;

@Exclude()
export class CommandStep extends LabeledStep {
    @Expose({ name: 'command' })
    @Transform((value: Command[]) => {
        if (!value || value.length === 0) {
            return undefined;
        }
        return value.length === 1
            ? value[0].command
            : value.map(c => c.command);
    })
    public readonly command: Command[] = [];

    @Expose()
    @Transform(transformKeyValueImpl)
    public readonly env: KeyValue<this>;

    private _id?: string;
    @Expose({ name: 'id' })
    private get id(): string | undefined {
        return this._id;
    }

    private _parallelism?: number;

    @Expose({ name: 'parallelism' })
    private get paralellism(): number | undefined {
        return this._parallelism;
    }

    @Expose()
    private concurrency?: number;

    @Expose({ name: 'concurrency_group' })
    private concurrencyGroup?: string;

    @Expose({ name: 'artifact_paths' })
    @Transform((paths: Set<string>) => (paths.size ? paths : undefined))
    private _artifactPaths: Set<string> = new Set();

    private _agents: Agents = new Map();

    @Expose()
    @Transform((agents: Agents) => (agents.size ? agents : undefined))
    get agents(): Agents {
        return this._agents;
    }

    private _timeout?: number;

    @Expose({ name: 'timeout_in_minutes' })
    get timeout(): number | undefined {
        if (this._timeout === Infinity || this._timeout === 0) {
            return undefined;
        } else if (this._timeout) {
            return this._timeout;
        }
        if (this.command) {
            const value = this.command.reduce((acc, command) => {
                acc += command.timeout;
                return acc;
            }, 0);
            if (value === Infinity || value === 0) {
                return undefined;
            } else if (value) {
                return value;
            }
        }
    }

    @Expose({ name: 'plugins' })
    @Transform(transformPlugins)
    public readonly plugins: Plugins<this> = new PluginsImpl(this);

    @Expose({ name: 'soft_fail' })
    @Transform((value: Set<ExitStatus>) => {
        if (!value.size) {
            return undefined;
        } else if (value.has('*')) {
            return true;
        } else {
            return [...value].map(s => ({
                // eslint-disable-next-line @typescript-eslint/camelcase
                exit_status: s,
            }));
        }
    })
    private _softFail: Set<ExitStatus> = new Set();

    @Expose({ name: 'skip' })
    @Transform((value: SkipValue | SkipFunction) => {
        if (typeof value === 'function') {
            value = value();
            assertSkipValue(value);
        }
        return value || undefined;
    })
    private _skip?: SkipValue | SkipFunction;

    @Expose()
    @Transform((value: RetryImpl<any>) =>
        value.hasValue() ? value : undefined,
    )
    public readonly retry: Retry<CommandStep> = new RetryImpl(this);

    constructor(plugin: Plugin, label?: string);
    constructor(command: string, label?: string);
    constructor(command: Command, label?: string);
    constructor(commands: (string | Plugin | Command)[], label?: string);
    constructor(
        command: string | Plugin | Command | (string | Plugin | Command)[],
        label?: string,
    ) {
        super();
        this.env = new KeyValueImpl(this);
        if (label) {
            this.withLabel(label);
        }
        if (Array.isArray(command)) {
            ow(command, ow.array.minLength(1));
            for (const c of command) {
                this.add(c);
            }
        } else {
            this.add(command);
        }
    }

    add(c: string | Plugin | Command): this {
        if (c instanceof Plugin) {
            this.plugins.add(c);
        } else if (typeof c === 'string') {
            this.command.push(new Command(c));
        } else if (c instanceof Command) {
            this.command.push(c);
        }
        return this;
    }

    /**
     *
     * @param timeout for the step in seconds.
     */
    withTimeout(timeout = Infinity): this {
        assertTimeout(timeout);
        this._timeout = timeout;
        return this;
    }

    withParallelism(parallelism: number): this {
        ow(parallelism, ow.number.integer.positive);
        this._parallelism = parallelism;
        return this;
    }

    withAgent(key: string, value: string): this {
        ow(key, ow.string.nonEmpty);
        ow(value, ow.string.nonEmpty);
        this._agents.set(key, value);
        return this;
    }

    withArtifactPath(...globs: string[]): this {
        ow(globs, ow.array.ofType(ow.string.nonEmpty));
        for (const glob of globs) {
            this._artifactPaths.add(glob);
        }
        return this;
    }

    withConcurrency(concurrency: number, group: string): this {
        ow(concurrency, ow.number.integer.positive);
        this.concurrency = concurrency;
        this.concurrencyGroup = group;
        return this;
    }

    withId(identifier: string): this {
        ow(identifier, ow.string.nonEmpty);
        this._id = identifier;
        return this;
    }

    withSoftFail(fail: ExitStatus | true): this {
        if (fail !== true) {
            ow(fail, exitStatusPredicate);
        }
        this._softFail.add(fail === true ? '*' : fail);
        return this;
    }

    skip(skip: SkipValue | SkipFunction): this {
        if (typeof skip !== 'function') {
            assertSkipValue(skip);
        }

        this._skip = skip;
        return this;
    }

    toString(): string {
        return (
            this.label ||
            (this.command
                ? `<${this.command.join(' && ') || '(empty)'}>`
                : this.plugins.toString())
        );
    }
}
