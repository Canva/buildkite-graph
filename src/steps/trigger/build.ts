import { Expose, Exclude, Transform } from 'class-transformer';
import { KeyValue, KeyValueImpl, transformKeyValueImpl } from '../../key_value';

export interface Build<T> {
    env: KeyValue<T>;
    metadata: KeyValue<T>;
    withMessage(message: string): T;
    withCommit(commit: string): T;
    withBranch(branch: string): T;
}
@Exclude()
export class BuildImpl<T> implements Build<T> {
    @Expose({ name: 'message' })
    private _message?: string;
    @Expose({ name: 'commit' })
    private _commit?: string;
    @Expose({ name: 'branch' })
    private _branch?: string;
    @Expose()
    @Transform(transformKeyValueImpl)
    public readonly env: KeyValue<T>;
    @Expose({ name: 'meta_data' })
    @Transform(transformKeyValueImpl)
    public readonly metadata: KeyValue<T>;
    constructor(private readonly triggerStep: T) {
        this.env = new KeyValueImpl(triggerStep);
        this.metadata = new KeyValueImpl(triggerStep);
    }
    withMessage(message: string): T {
        this._message = message;
        return this.triggerStep;
    }
    withCommit(commit: string): T {
        this._commit = commit;
        return this.triggerStep;
    }
    withBranch(branch: string): T {
        this._branch = branch;
        return this.triggerStep;
    }
    hasData() {
        return (
            this._branch ||
            this._commit ||
            typeof this._message !== 'undefined' ||
            (this.env as KeyValueImpl<T>).vars.size ||
            (this.metadata as KeyValueImpl<T>).vars.size
        );
    }
}
