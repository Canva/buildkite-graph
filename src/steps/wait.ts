import { BaseStep } from '../base';
import { Serializable } from '../index';

export class WaitStep implements BaseStep, Serializable {
  constructor(public continueOnFailure = false) {}

  toString(): string {
    /* istanbul ignore next */
    return this.continueOnFailure ? '[wait; continues on failure]' : '[wait]';
  }

  async toJson(): Promise<Record<string, unknown>> {
    return {
      wait: null,
      continue_on_failure: this.continueOnFailure || undefined,
    };
  }
}
