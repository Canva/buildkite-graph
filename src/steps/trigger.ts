import { Exclude, Expose, Transform } from 'class-transformer';
import { Pipeline } from '../';
import { LabeledStep } from '../base';
import { Build, BuildImpl } from './trigger/build';

@Exclude()
export class TriggerStep extends LabeledStep {
    @Expose()
    get trigger(): string {
        return this._trigger instanceof Pipeline
            ? this._trigger.slug()
            : this._trigger;
    }

    @Expose({ name: 'async' })
    @Transform((value: boolean) => (value ? value : undefined))
    private _async = false;

    @Expose()
    @Transform((value: BuildImpl<any>) => (value.hasData() ? value : undefined))
    public readonly build: Build<TriggerStep> = new BuildImpl(this);

    constructor(
        private readonly _trigger: Pipeline | string,
        label?: string,
        async = false,
    ) {
        super();
        if (label) {
            this.withLabel(label);
        }
        this._async = async;
    }
    async(async: boolean): this {
        this._async = async;
        return this;
    }
    toString(): string {
        return this.label || `[trigger ${this.trigger}]`;
    }
}
