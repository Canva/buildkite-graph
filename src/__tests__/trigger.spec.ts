import { Pipeline } from '../';
import { TriggerStep } from '../steps/trigger';
import { createTest } from './helpers';

describe('buildkite-graph', () => {
    describe('Steps', () => {
        describe('Trigger', () => {
            createTest('can trigger another pipeline', () => [
                new Pipeline('whatever').add(
                    new TriggerStep(new Pipeline('another-build')),
                ),
                new Pipeline('whatever').add(new TriggerStep('another-build')),
            ]);

            createTest('with a label', () => [
                new Pipeline('whatever').add(
                    new TriggerStep('another', 'Trigger another pipeline'),
                ),
            ]);

            createTest('limit branches', () => [
                new Pipeline('whatever').add(
                    new TriggerStep('another').withBranch('master'),
                ),
            ]);
            createTest('async', () => [
                new Pipeline('whatever').add(
                    new TriggerStep('another', 'Label', true),
                ),
                new Pipeline('whatever').add(
                    new TriggerStep('another', 'Label', true).async(false),
                ),
                new Pipeline('whatever').add(
                    new TriggerStep('another').async(true),
                ),
            ]);

            createTest('build', () => [
                new Pipeline('whatever').add(
                    new TriggerStep('another').build.env
                        .set('KEY', 'VALUE')
                        .build.env.set('ANOTHER_KEY', 'VALUE'),
                ),
                new Pipeline('whatever').add(
                    new TriggerStep('another').build.withBranch('rease-bla'),
                ),
                new Pipeline('whatever').add(
                    new TriggerStep('another').build.withCommit('c0ffee'),
                ),
                new Pipeline('whatever').add(
                    new TriggerStep('another').build.withMessage('My messge'),
                ),
                new Pipeline('whatever').add(
                    new TriggerStep('another').build.metadata
                        .set('release-version', '1.1')
                        .build.metadata.set('some-other', 'value'),
                ),
            ]);
        });
    });
});
