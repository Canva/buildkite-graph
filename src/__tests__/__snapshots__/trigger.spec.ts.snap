// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildkite-graph Steps Trigger async dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"Label\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"Label\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger async dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"Label\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"Label\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger async dot 3`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger async json 1`] = `
Object {
  "steps": Array [
    Object {
      "async": true,
      "label": "Label",
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger async json 2`] = `
Object {
  "steps": Array [
    Object {
      "label": "Label",
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger async json 3`] = `
Object {
  "steps": Array [
    Object {
      "async": true,
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger async yaml 1`] = `
"steps:
  - label: Label
    trigger: another
    async: true
"
`;

exports[`buildkite-graph Steps Trigger async yaml 2`] = `
"steps:
  - label: Label
    trigger: another
"
`;

exports[`buildkite-graph Steps Trigger async yaml 3`] = `
"steps:
  - trigger: another
    async: true
"
`;

exports[`buildkite-graph Steps Trigger build dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger build dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger build dot 3`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger build dot 4`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger build dot 5`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger build json 1`] = `
Object {
  "steps": Array [
    Object {
      "build": Object {
        "env": Object {
          "ANOTHER_KEY": "VALUE",
          "KEY": "VALUE",
        },
      },
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger build json 2`] = `
Object {
  "steps": Array [
    Object {
      "build": Object {
        "branch": "rease-bla",
      },
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger build json 3`] = `
Object {
  "steps": Array [
    Object {
      "build": Object {
        "commit": "c0ffee",
      },
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger build json 4`] = `
Object {
  "steps": Array [
    Object {
      "build": Object {
        "message": "My messge",
      },
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger build json 5`] = `
Object {
  "steps": Array [
    Object {
      "build": Object {
        "meta_data": Object {
          "release-version": "1.1",
          "some-other": "value",
        },
      },
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger build yaml 1`] = `
"steps:
  - trigger: another
    build:
      env:
        KEY: VALUE
        ANOTHER_KEY: VALUE
"
`;

exports[`buildkite-graph Steps Trigger build yaml 2`] = `
"steps:
  - trigger: another
    build:
      branch: rease-bla
"
`;

exports[`buildkite-graph Steps Trigger build yaml 3`] = `
"steps:
  - trigger: another
    build:
      commit: c0ffee
"
`;

exports[`buildkite-graph Steps Trigger build yaml 4`] = `
"steps:
  - trigger: another
    build:
      message: My messge
"
`;

exports[`buildkite-graph Steps Trigger build yaml 5`] = `
"steps:
  - trigger: another
    build:
      meta_data:
        release-version: '1.1'
        some-other: value
"
`;

exports[`buildkite-graph Steps Trigger can trigger another pipeline dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another-build]\\" [ color = \\"grey\\" ];
}

  \\"another-build\\" [ shape =Msquare ];
  \\"[trigger another-build]\\" -> \\"another-build\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger can trigger another pipeline dot 2`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another-build]\\" [ color = \\"grey\\" ];
}

  \\"another-build\\" [ shape =Msquare ];
  \\"[trigger another-build]\\" -> \\"another-build\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger can trigger another pipeline json 1`] = `
Object {
  "steps": Array [
    Object {
      "trigger": "another-build",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger can trigger another pipeline json 2`] = `
Object {
  "steps": Array [
    Object {
      "trigger": "another-build",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger can trigger another pipeline yaml 1`] = `
"steps:
  - trigger: another-build
"
`;

exports[`buildkite-graph Steps Trigger can trigger another pipeline yaml 2`] = `
"steps:
  - trigger: another-build
"
`;

exports[`buildkite-graph Steps Trigger limit branches dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"[trigger another]\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"[trigger another]\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger limit branches json 1`] = `
Object {
  "steps": Array [
    Object {
      "branches": "master",
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger limit branches yaml 1`] = `
"steps:
  - branches: master
    trigger: another
"
`;

exports[`buildkite-graph Steps Trigger with a label dot 1`] = `
"digraph \\"whatever\\" {
  graph [ compound =true ];
subgraph cluster_0 {
  graph [ color = \\"black\\" ];
  \\"Trigger another pipeline\\" [ color = \\"grey\\" ];
}

  \\"another\\" [ shape =Msquare ];
  \\"Trigger another pipeline\\" -> \\"another\\" [ label = \\"triggers\\" ];
}
"
`;

exports[`buildkite-graph Steps Trigger with a label json 1`] = `
Object {
  "steps": Array [
    Object {
      "label": "Trigger another pipeline",
      "trigger": "another",
    },
  ],
}
`;

exports[`buildkite-graph Steps Trigger with a label yaml 1`] = `
"steps:
  - label: Trigger another pipeline
    trigger: another
"
`;
